File: python/virtualcan/can_message.py
"""Can message class."""

import struct
import binascii


class CanMessage:
    def __init__(self, id, extended, data):
        self.id = id
        self.extended = extended
        self.data = data

    def __repr__(self):
        hex_data = " ".join(f"{b:02X}" for b in self.data)
        return f"Can message id={self.id} extended={self.extended} data={hex_data}"

    def __eq__(self, other):
        return (self.id, self.extended, self.data) == (
            other.id,
            other.extended,
            other.data,
        )

    def to_json(self):
        return {
            "id": int(self.id),
            "extended": int(self.extended),
            "bindata": binascii.hexlify(self.data).decode("ascii"),
        }

    @classmethod
    def from_json(cls, json_data):
        id = json_data["id"]
        extended = bool(json_data["extended"])
        data = binascii.unhexlify(json_data["bindata"])
        return cls(id, extended, data)

    _CAN_FMT = ">IBB8s"
    _CAN_FLAGS_EXTENDED = 1

    def to_bytes(self):
        """Serialize this can message to some binary format."""
        flags = 0
        if self.extended:
            flags |= self._CAN_FLAGS_EXTENDED

        data_len = len(self.data)
        bindata = struct.pack(self._CAN_FMT, self.id, flags, data_len, self.data)
        return bindata

    @classmethod
    def from_bytes(cls, bindata):
        """Create a can message from some binary data."""
        can_id, flags, data_len, data = struct.unpack(cls._CAN_FMT, bindata)
        extended = bool(flags & cls._CAN_FLAGS_EXTENDED)
        data = data[:data_len]
        return cls(can_id, extended, data)

File: cpp/virtualcan/can_message.cpp

#include "can_message.h"
#include "util.h"
#include <stdio.h>

namespace virtualcan {

void print_can_message(const CanMessage* can_msg)
{
    int i;
    printf("Can msg: id=%X data=", can_msg->id);
    for (i=0; i<can_msg->data_size; i++)
    {
        printf("%02X ", can_msg->data[i]);
    }

    printf("\n");
}

// serialize a can message in some format
void pack_can_msg(uint8_t* buffer, CanMessage* msg)
{
    pack_u32(buffer, msg->id);
    uint8_t flags = 0;
    if (msg->extended)
    {
        flags |= 1;
    }
    buffer[4] = flags;
    buffer[5] = msg->data_size;
    int i;
    for (i=0;i< msg->data_size;i++)
    {
        buffer[6 + i] = msg->data[i];
    }
}

void unpack_can_msg(uint8_t* buffer, CanMessage* msg)
{
    msg->id = unpack_u32(buffer);
    uint8_t flags = buffer[4];
    msg->extended = (flags & 1) ? 1 : 0;
    msg->data_size = buffer[5];

    int i;
    for (i=0;i< msg->data_size;i++)
    {
        msg->data[i] = buffer[6 + i];
    }
}

}
